fdbserver/workloads/MetaclusterRestoreWorkload.actor.cpp

```diff
@@ -578,14 +603,69 @@ struct MetaclusterRestoreWorkload : TestWorkload {
 						messages.clear();
 					}
 
-					wait(MetaclusterAPI::restoreCluster(
-					    self->managementDb,
-					    clusterItr->first,
-					    clusterItr->second.db->getConnectionRecord()->getConnectionString(),
-					    ApplyManagementClusterUpdates::False,
-					    RestoreDryRun::False,
-					    ForceJoin(deterministicRandom()->coinflip()),
-					    &messages));
+					state int numRestores = deterministicRandom()->randomInt(1, 3);
+					state bool successfulRestore = false;
+					loop {
+						state std::vector<std::vector<std::string>> messagesList(numRestores);
+						state std::vector<Future<ErrorOr<Void>>> restoreFutures;
+						for (; numRestores > 0; numRestores--) {
+							restoreFutures.push_back(errorOr(MetaclusterAPI::restoreCluster(
+							    self->managementDb,
+							    clusterItr->first,
+							    clusterItr->second.db->getConnectionRecord()->getConnectionString(),
+							    ApplyManagementClusterUpdates::False,
+							    RestoreDryRun::False,
+							    ForceJoin(deterministicRandom()->coinflip()),
+							    &messagesList[restoreFutures.size()])));
+							wait(delay(deterministicRandom()->random01() * 5));
+						}
+
+						wait(waitForAll(restoreFutures));
+
+						Optional<Error> nonConflictError;
+						for (int i = 0; i < restoreFutures.size(); ++i) {
+							ErrorOr<Void> result = restoreFutures[i].get();
+							fmt::print("Restore result for {}: {}\n",
+							           printable(clusterItr->first),
+							           result.isError() ? result.getError().what() : "success");
+							if (!result.isError()) {
+								ASSERT(!successfulRestore);
+								successfulRestore = true;
+								messages = messagesList[i];
+							} else if (result.getError().code() != error_code_conflicting_restore &&
+							           result.getError().code() != error_code_cluster_already_registered &&
+							           result.getError().code() != error_code_cluster_already_exists &&
+							           !nonConflictError.present()) {
+								nonConflictError = result.getError().code();
+								messages = messagesList[i];
+							}
+						}
+
+						if (nonConflictError.present()) {
+							ASSERT(!successfulRestore);
```

Comment:
This might be me just misunderstanding how this test works: If one restore future succeeds, and a different one fails with a `nonConflictError`, that's a state that we cannot be in or recover from? Isn't it possible to get `operation_failed` or timeout errors?

Reply:
The `nonConflictError` variable is meant to capture whether the restore failed with an error related to two of them running simultaneously. If there is some other problem that is causing errors, it should be causing the error for both of them (or at least that's the theory), so we wouldn't expect one to succeed and the other to not.

We won't get any timeouts from this because we aren't setting a timeout on the restore.